Should we add the `-Werror` compiler flag to the CI builds?

We are very close to eliminate all the compilation warnings for WarpX*, at least when `g++` is used (`clang` is more severe).
As @EZoni and @ax3l have suggested in the past, should we add the `-Werror` compiler flag to the CI builds, so that we can be sure that new PRs do not introduce new warnings?

* to give and idea, if I compile with `cmake` with the default settings and `g++ 9.3.0` I have zero warnings. I get basically just one warning with `g++ 10.2.0`

We are close! :rocket:

I currently still see those warnings on Summit:
```
./Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp: In member function 'void WarpXLaserProfiles::FromTXYEFileLaserProfile::parse_txye_file(std::__cxx11::string)':
./Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:134:58: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
         m_params.nt = reinterpret_cast<uint32_t*>(buf)[0];
                                                          ^
./Source/Parallelization/WarpXRegrid.cpp: In member function 'void WarpX::LoadBalance()':
./Source/Parallelization/WarpXRegrid.cpp:80:19: warning: conversion to 'std::vector<int, std::allocator<int> >::size_type {aka long unsigned int}' from 'amrex::Real {aka double}' may alter its value [-Wfloat-conversion]
                 pmap.resize(nboxes);
```

The 2nd is #1453

In CI I see in AMReX:
- https://github.com/AMReX-Codes/amrex/issues/1478
- https://github.com/AMReX-Codes/amrex/issues/1479

> We are close! rocket
> 
> I currently still see those warnings on Summit:
> 
> ```
> ./Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp: In member function 'void WarpXLaserProfiles::FromTXYEFileLaserProfile::parse_txye_file(std::__cxx11::string)':
> ./Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:134:58: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
>          m_params.nt = reinterpret_cast<uint32_t*>(buf)[0];
>                                                           ^
> ./Source/Parallelization/WarpXRegrid.cpp: In member function 'void WarpX::LoadBalance()':
> ./Source/Parallelization/WarpXRegrid.cpp:80:19: warning: conversion to 'std::vector<int, std::allocator<int> >::size_type {aka long unsigned int}' from 'amrex::Real {aka double}' may alter its value [-Wfloat-conversion]
>                  pmap.resize(nboxes);
> ```
> 
> The 2nd is #1453
> 
> In CI I see in AMReX:
> 
>     * [AMReX-Codes/amrex#1478](https://github.com/AMReX-Codes/amrex/issues/1478)
> 
>     * [AMReX-Codes/amrex#1479](https://github.com/AMReX-Codes/amrex/issues/1479)

Uhm... it's UB right? I think that it is my fault.  I can fix it with something like `std::memcpy(&m_params.nt, buf, sizeof(uint32_t))` .  Actually, I used this unsafe type punning technique multiple times in that file. I will open a PR.

Btw, which combination of compiler and compiler options gave you this warning? I have never seen it on my machines...

> ```
> ./Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp: In member function 'void WarpXLaserProfiles::FromTXYEFileLaserProfile::parse_txye_file(std::__cxx11::string)':
> ./Source/Laser/LaserProfilesImpl/LaserProfileFromTXYEFile.cpp:134:58: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
>          m_params.nt = reinterpret_cast<uint32_t*>(buf)[0];

This PR should fix the issue: https://github.com/ECP-WarpX/WarpX/pull/1470 .

--- a/.github/workflows/cuda.yml
+++ b/.github/workflows/cuda.yml
@@ -10,6 +10,7 @@ jobs:
   build_nvcc:
     name: NVCC 11.0.2 SP
     runs-on: ubuntu-18.04
+    env: {CXXFLAGS: "-Werror"}
     steps:
     - uses: actions/checkout@v2
     - name: install dependencies

--- a/.github/workflows/hip.yml
+++ b/.github/workflows/hip.yml
@@ -6,6 +6,7 @@ jobs:
   build_hip:
     name: HIP SP
     runs-on: ubuntu-20.04
+    env: {CXXFLAGS: "-Werror -Wno-deprecated-declarations"}
     steps:
     - uses: actions/checkout@v2
     - name: install dependencies

--- a/.github/workflows/intel.yml
+++ b/.github/workflows/intel.yml
@@ -58,6 +58,7 @@ jobs:
   build_icpx:
     name: oneAPI ICX SP
     runs-on: ubuntu-20.04
+    env: {CXXFLAGS: "-Werror -Wno-error=pass-failed"}
     steps:
     - uses: actions/checkout@v2
     - name: install dependencies
@@ -97,6 +98,7 @@ jobs:
   build_dpcc:
     name: oneAPI DPC++ SP
     runs-on: ubuntu-20.04
+    env: {CXXFLAGS: "-Werror"}
     steps:
     - uses: actions/checkout@v2
     - name: install dependencies

--- a/.github/workflows/macos.yml
+++ b/.github/workflows/macos.yml
@@ -6,6 +6,7 @@ jobs:
   build_gcc9:
     name: AppleClang
     runs-on: macos-latest
+    env: {CXXFLAGS: "-Werror -Wno-error=pass-failed"}
     steps:
     - uses: actions/checkout@v2
     - name: install dependencies

--- a/.github/workflows/ubuntu.yml
+++ b/.github/workflows/ubuntu.yml
@@ -43,6 +43,7 @@ jobs:
         export WarpX_QED_TABLE_GEN=ON
         export CC=$(which clang)
         export CXX=$(which clang++)
+        export CXXFLAGS="-Werror -Wno-error=pass-failed"
         python3 -m pip install -v .
 
     - name: run pywarpx

